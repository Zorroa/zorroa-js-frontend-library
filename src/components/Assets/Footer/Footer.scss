@import '../../../styles/core-mixins.scss';

.AssetsFooter {
  align-self: center;
  z-index: 1;
  margin-top: $zorroa-metric-base * -6;
  margin-bottom: $zorroa-metric-base;
  max-width: $zorroa-metric-base * 50;
  border-radius: 3px;
  box-shadow: 0 4px 6px rgba($zorroa-color-black, 0.2);
  color: $zorroa-color-gray-dark-4;
  background-color: rgba(0,0,0,0.8);
  padding: 0 $zorroa-metric-base;

  width: auto;
  height: $zorroa-metric-base * 4;

  @include flexRow;
  @include flexAlignItemsCenter;
  @include flexJustifySpaceBetween;

  > * {
      height: 100%;
  }

  button:disabled {
    color: $zorroa-darkgray;
    &:hover {
      color: $zorroa-darkgray;
    }
  }

  .Workspace.embedMode & {
    display: none;
  }

  &__icon {
    color: $zorroa-color-gray-dark-4;
    width: $zorroa-metric-base * 3;
    height: $zorroa-metric-base * 3;
    border: 0;
    outline: 0;
    font-size: 20px;
    cursor: pointer;
    border-radius: 3px;
    @include flexRowCenter;

    &:focus,
    &:hover {
      color: $zorroa-color-green-1;
    }

    &--enabled {
      color: $zorroa-color-green-1;
      background-color: rgba($zorroa-color-black, 0.75);

      &:focus,
      &:hover {
        color: darken($zorroa-color-green-1, 15%);
      }
    }
  }
}
