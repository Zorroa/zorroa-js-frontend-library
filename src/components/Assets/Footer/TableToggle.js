import PropTypes from 'prop-types'
import React from 'react'
import classnames from 'classnames'

const TableToggle = props => (
  <div className="TableToggle" title="Show metadata spreadsheet">
    <div
      onClick={props.onClick}
      className={classnames(
        'TableToggle-button AssetsFooter__icon',
        'icon-table',
        {
          'AssetsFooter__icon--enabled': props.enabled,
        },
      )}
    />
  </div>
)

TableToggle.propTypes = {
  onClick: PropTypes.func.isRequired,
  enabled: PropTypes.bool.isRequired,
}

export default TableToggle
