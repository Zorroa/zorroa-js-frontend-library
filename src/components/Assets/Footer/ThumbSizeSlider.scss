@import '../../../styles/core-mixins.scss';

.ThumbSizeSlider {
  padding: 0 $zorroa-metric-base;
  border-radius: 3px;
}

.ThumbSizeSlider-slider {
  position: relative;
  padding: 0 $zorroa-metric-base / 2;
  margin: 0;
  background: none;
}

.ThumbSizeSlider-smaller {
  width: $zorroa-metric-base * 3;
  height: $zorroa-metric-base * 3;
}

.ThumbSizeSlider-bigger {
  [class*='icon-'] {
    font-size: 12px;
  }
}
.ThumbSizeSlider-box {
  position: relative;
  input::-webkit-slider-thumb {
    border: none;
    background: $zorroa-lightgray;
    &:active { background-color: $zorroa-sickly-green; }
  }
  input::-moz-range-thumb {
    border: none;
    background: $zorroa-lightgray;
    &:active { background-color: $zorroa-sickly-green; }
  }
}
.ThumbSizeSlider-line-box {
  position: absolute;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  pointer-events: none;
}
.ThumbSizeSlider-line {
  width: $zorroa-metric-base * 12;
  height: 1px;
  background-color: $zorroa-gray;
  &:active { background-color: $zorroa-sickly-green; }
}
