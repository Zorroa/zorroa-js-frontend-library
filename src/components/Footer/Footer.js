import classnames from 'classnames'

import Sidebar from '../Sidebar'

export default class Footer extends Sidebar {
  static displayName () {
    return 'Footer'
  }

  buttonChar () {
    // Select the right or left facing triangle unicode char using XOR
    return this.state.open !== this.props.isRightEdge ? '\u25BC' : '\u25B2'
  }

  buttonClassNames () {
    return 'footer-button'
  }

  sidebarClassNames () {
    return classnames('footer', {
      'footer-open': this.state.open
    })
  }
}
