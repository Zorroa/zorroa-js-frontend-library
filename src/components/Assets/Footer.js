import React, { PropTypes } from 'react'

import AssetCounter from './AssetCounter'
import ThumbSizeSlider from './ThumbSizeSlider'
import ThumbLayoutSelector from './ThumbLayoutSelector'
import TableToggle from './TableToggle'

const Footer = (props) => (
  <div className="assets-footer flexRow flexJustifySpaceBetween flexAlignItemsCenter">
    <div className="flexRow">
      <AssetCounter loaded={props.loaded} total={props.total}/>
    </div>
    <div className="flexRow flexJustifyEnd flexAlignItemsCenter">
      <ThumbSizeSlider value={props.thumbSize} onChange={props.handleThumbSize} />
      <ThumbLayoutSelector thumbLayout={props.layout} onClick={props.handleLayout} />
      <TableToggle enabled={props.showTable} onClick={props.toggleShowTable} />
    </div>
  </div>
)

Footer.propTypes = {
  loaded: PropTypes.number.isRequired,
  total: PropTypes.number.isRequired,
  showTable: PropTypes.bool.isRequired,
  toggleShowTable: PropTypes.func.isRequired,
  layout: PropTypes.string.isRequired,
  handleLayout: PropTypes.func.isRequired,
  thumbSize: PropTypes.number.isRequired,
  handleThumbSize: PropTypes.func.isRequired
}

export default Footer
