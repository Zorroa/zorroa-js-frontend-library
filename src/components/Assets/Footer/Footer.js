import PropTypes from 'prop-types'
import React, { PureComponent } from 'react'

import ThumbSizeSlider from './ThumbSizeSlider'
import ThumbLayoutSelector from './ThumbLayoutSelector'
import TableToggle from './TableToggle'
import User from '../../../models/User'

export default class Footer extends PureComponent {
  static propTypes = {
    showTable: PropTypes.bool.isRequired,
    layout: PropTypes.string.isRequired,
    handleLayout: PropTypes.func,
    thumbSize: PropTypes.number.isRequired,
    user: PropTypes.instanceOf(User),
    userSettings: PropTypes.object.isRequired,
    actions: PropTypes.shape({
      setThumbSize: PropTypes.func.isRequired,
      setThumbLayout: PropTypes.func.isRequired,
      showTable: PropTypes.func.isRequired,
      saveUserSettings: PropTypes.func.isRequired,
    }),
  }

  toggleShowTable = () => {
    const { showTable, user, userSettings, actions } = this.props
    actions.showTable(!showTable)
    actions.saveUserSettings(user, { ...userSettings, showTable: !showTable })
  }

  handleLayout() {
    if (typeof this.props.handleLayout === 'function') {
      this.props.handleLayout()
    }
  }

  changeLayout = layout => {
    if (this.props.layout !== layout) {
      this.props.actions.setThumbLayout(layout)
      this.props.actions.saveUserSettings(this.props.user, {
        ...this.props.userSettings,
        thumbLayout: layout,
      })
      this.handleLayout()
    }
  }

  changeThumbSize = thumbSize => {
    if (typeof thumbSize !== 'number') {
      return
    }

    if (this.props.thumbSize !== thumbSize) {
      this.props.actions.setThumbSize(thumbSize)
      this.props.actions.saveUserSettings(this.props.user, {
        ...this.props.userSettings,
        thumbSize,
      })
      this.handleLayout()
    }
  }

  render() {
    return (
      <div className="AssetsFooter">
        <ThumbSizeSlider
          value={this.props.thumbSize}
          onChange={this.changeThumbSize}
        />
        <ThumbLayoutSelector
          thumbLayout={this.props.layout}
          onClick={this.changeLayout}
        />
        <TableToggle
          enabled={this.props.showTable}
          onClick={this.toggleShowTable}
        />
      </div>
    )
  }
}
