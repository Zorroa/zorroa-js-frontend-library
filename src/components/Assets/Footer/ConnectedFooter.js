import { bindActionCreators } from 'redux'
import { connect } from 'react-redux'
import { withRouter } from 'react-router-dom'
import { saveUserSettings } from '../../../actions/authAction'
import {
  setThumbSize,
  setThumbLayout,
  showTable,
} from '../../../actions/appActions'
import Footer from './Footer'

const ConnectedFooter = connect(
  state => ({
    uxLevel: state.app.uxLevel,
    user: state.auth.user,
    userSettings: state.app.userSettings,
    thumbSize: state.app.thumbSize,
    layout: state.app.thumbLayout,
    showTable: state.app.showTable,
    showQuickview: state.app.showQuickview,
  }),
  dispatch => ({
    actions: bindActionCreators(
      {
        setThumbSize,
        setThumbLayout,
        showTable,
        saveUserSettings,
      },
      dispatch,
    ),
  }),
)(Footer)

export default withRouter(ConnectedFooter)
