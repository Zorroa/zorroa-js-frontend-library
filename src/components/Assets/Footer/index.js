import ConnectedFooter from './ConnectedFooter'
import './Footer.scss'
import './ThumbSizeSlider.scss'
export default ConnectedFooter
