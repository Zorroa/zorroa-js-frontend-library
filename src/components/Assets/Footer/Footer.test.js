/* eslint-env jest */

import React from 'react'
import { shallow, configure } from 'enzyme'
import Adapter from 'enzyme-adapter-react-15'
import renderer from 'react-test-renderer'
import Footer from './Footer'
import User from '../../../models/User'

configure({ adapter: new Adapter() })

function generateActions() {
  const setThumbSize = jest.fn()
  const setThumbLayout = jest.fn()
  const showTable = jest.fn()
  const saveUserSettings = jest.fn()
  const actions = {
    setThumbSize,
    setThumbLayout,
    showTable,
    saveUserSettings,
  }

  return actions
}

function generateProps(customProps) {
  return {
    actions: generateActions(),
    showTable: true,
    layout: 'masonry',
    handleLayout: jest.fn(),
    thumbSize: 200,
    user: new User({ id: 'b' }),
    userSettings: {},
    ...customProps,
  }
}

describe('<Footer />', () => {
  describe('toggleShowTable()', () => {
    it('Should toggle the table to the oposite of the current value', () => {
      const props = generateProps()
      const component = shallow(<Footer {...props} />)
      component.instance().toggleShowTable()
      const showTable = props.actions.showTable.mock.calls[0][0]
      expect(showTable).toBe(false)
    })
  })

  describe('handleLayout()', () => {
    it('Should not crash when no handler is defined', () => {
      const props = generateProps()
      props.handleLayout = undefined
      const component = shallow(<Footer {...props} />)
      component.instance().handleLayout()
    })

    it('Should call the handler when it is defined', () => {
      const props = generateProps()
      const handleLayout = props.handleLayout
      const component = shallow(<Footer {...props} />)
      component.instance().handleLayout()
      expect(handleLayout.mock.calls.length).toBe(1)
    })
  })

  describe('changeLayout()', () => {
    it('Should not call `handleLayout()` if the layout name the same', () => {
      const props = generateProps()
      const component = shallow(<Footer {...props} />)
      component.instance().changeLayout('masonry')
      expect(props.handleLayout.mock.calls.length).toBe(0)
    })

    it('Should call `handleLayout()` when the layout name is different', () => {
      const props = generateProps()
      const handleLayout = props.handleLayout
      const component = shallow(<Footer {...props} />)
      component.instance().changeLayout('grid')
      expect(handleLayout.mock.calls.length).toBe(1)
    })
  })

  describe('changeThumbSize()', () => {
    it('Should not call `handleLayout()` if the thumbnail size is the same', () => {
      const props = generateProps()
      const component = shallow(<Footer {...props} />)
      component.instance().changeThumbSize(200)
      expect(props.handleLayout.mock.calls.length).toBe(0)
    })

    it('Should not call `handleLayout()` if the thumbnail size is not a number', () => {
      const props = generateProps()
      const component = shallow(<Footer {...props} />)
      component.instance().changeThumbSize('200')
      expect(props.handleLayout.mock.calls.length).toBe(0)
    })

    it('Should call `handleLayout()` when the thumbnail size is different', () => {
      const props = generateProps()
      const handleLayout = props.handleLayout
      const component = shallow(<Footer {...props} />)
      component.instance().changeThumbSize(300)
      expect(handleLayout.mock.calls.length).toBe(1)
    })
  })

  describe('render()', () => {
    const props = generateProps()
    const tree = renderer.create(<Footer {...props} />).toJSON()
    expect(tree).toMatchSnapshot()
  })
})
