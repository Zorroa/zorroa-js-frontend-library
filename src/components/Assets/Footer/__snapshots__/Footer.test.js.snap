// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
<div
  className="AssetsFooter"
>
  <div
    className="ThumbSizeSlider flexRow flexAlignItemsCenter"
  >
    <div
      className="ThumbSizeSlider-smaller AssetsFooter__icon flexOff flexRowCenter"
      onClick={[Function]}
      title="Smaller thumbnails"
    >
      <i
        className="icon-picture2"
        style={
          Object {
            "fontSize": "12px",
          }
        }
      />
    </div>
    <div
      className="ThumbSizeSlider-box flexRowCenter"
    >
      <div
        className="ThumbSizeSlider-line-box flexRowCenter"
      >
        <div
          className="ThumbSizeSlider-line"
        />
      </div>
      <input
        className="ThumbSizeSlider-slider"
        max="341.3333333333333"
        min="146.28571428571428"
        onChange={[Function]}
        step="0.5"
        title="Thumbnail size"
        type="range"
        value={200}
      />
    </div>
    <div
      className="ThumbSizeSlider-bigger Footer__icon flexOff flexRowCenter"
      onClick={[Function]}
      title="Larger thumbnails"
    >
      <i
        className="icon-picture2"
        style={
          Object {
            "fontSize": "20px",
          }
        }
      />
    </div>
  </div>
  <div
    className="thumb-layout"
  >
    <div
      className="AssetsFooter__icon icon-layout2"
      onClick={[Function]}
      title="Grid layout"
    />
    <div
      className="AssetsFooter__icon icon-layout3 AssetsFooter__icon--enabled"
      onClick={[Function]}
      title="Flow layout"
    />
  </div>
  <div
    className="TableToggle"
    title="Show metadata spreadsheet"
  >
    <div
      className="TableToggle-button AssetsFooter__icon icon-table AssetsFooter__icon--enabled"
      onClick={[Function]}
    />
  </div>
</div>
`;
